flight/apps/adcs/mcm.py

```diff
@@ -1,10 +1,140 @@
-# Attitude Determination and Control Subsystem (ADCS)
-# Magnetic Control Module
+"""
+Attitude Determination and Control Subsystem (ADCS)
+Magnetic Control Module
+"""
 
+import copy
+from typing import Tuple
 
-# TODO: implement controllers: B-dot, Bcross, Sun pointing
-# TODO: implement desired moment allocation to voltages
-# ...
+from apps.adcs.consts import MagnetorquerConst, MCMConst, ModeConst, PhysicalConst
+from apps.adcs.math import is_near
+from hal.configuration import SATELLITE
+from ulab import numpy as np
 
-# To apply the voltages to the coils, the following function is used:
-# SATELLITE.APPLY_MAGNETIC_CONTROL({"XP": 5.0, "XM": 5.1, "YP": 4.8, "YM": 5.0, "ZP": 4.1, "ZM": 4.1})
+
+def get_spin_stabilizing_dipole_moment(
+    magnetic_field: np.ndarray,
+    spin_error: np.ndarray,
+) -> np.ndarray:
+    """
+    B-cross law: https://arc.aiaa.org/doi/epdf/10.2514/1.53074.
+    Augmented with tanh function for soft clipping.
+    All sensor estimates are in the body-fixed reference frame.
+    """
+    u_dir = MCMConst.BCROSS_GAIN * np.cross(magnetic_field, spin_error)
+    return ControllerHandler.u_max * np.tanh(u_dir)
+
+
+def get_sun_pointing_dipole_moment(
+    magnetic_field: np.ndarray,
+    pointing_error: np.ndarray,
+) -> np.ndarray:
+    """
+    Bang-bang sun pointing slewing law.
+    All sensor estimates are in the body-fixed reference frame.
+    """
+    u_dir = np.cross(magnetic_field, pointing_error)
+    u_dir_norm = np.linalg.norm(u_dir)
+
+    # Check for zeros to prevent division by zero
+    if is_near(u_dir_norm, 0.0):
+        return np.zeros(3)
+    else:
+        return ControllerHandler.u_max * u_dir / u_dir_norm
+
+
+class ControllerHandler:
+    """
+    "Handler" for determining controller references.
+    """
+
+    # Max dipole moment computed from max voltage
+    u_max = np.zeros(3)
+
+    # Init spin direction
+    _eigvals, _eigvecs = np.linalg.eig(PhysicalConst.INERTIA_MAT)
+    _unscaled_axis = _eigvecs[:, np.argmax(_eigvals)]
+    spin_axis = _unscaled_axis / np.linalg.norm(_unscaled_axis)
+    if spin_axis[np.argmax(np.abs(spin_axis))] < 0:
+        spin_axis = -spin_axis
+
+    # References and targets
+    ang_vel_reference = spin_axis * MCMConst.REF_FACTOR * ModeConst.STABLE_TOL
+    ang_vel_target = np.linalg.norm(ang_vel_reference)
+
+    @classmethod
+    def update_max_dipole_moment(self) -> None:
+        u_max = np.zeros(3)
+        for row in MagneticCoilAllocator.mat:
+            if not np.all(row == 0.0):
+                row_norm = np.linalg.norm(row)
+                u_max += row / row_norm * (MagnetorquerConst.V_MAX / MagnetorquerConst.V_CONVERT)
+        self.u_max = u_max
```

Comment:
I think there's a more efficient way of doing this part than looping through the matrix, but I haven't quite figured out what it is.

Reply:
one thing already is to precompute (MagnetorquerConst.V_MAX / MagnetorquerConst.V_CONVERT) into a multiplicative constant (division are extremely expensive)

Reply:
precompute (MagnetorquerConst.V_MAX / MagnetorquerConst.V_CONVERT) into a multiplicative constant, divisions are expensive

Reply:
might be worthwile for us to measure the processing speed of this (using the nanosec clock)